network.go

```diff
@@ -0,0 +1,511 @@
+	// VM will be used instead.
+	ContainerID string
+
+	// BinPath (optional) is a list of directories in which CNI plugin binaries
+	// will be sought. If not provided, defaults to just "/opt/bin/CNI"
+	BinPath []string
+
+	// ConfDir (optional) is the directory in which CNI configuration files
+	// will be sought. If not provided, defaults to "/etc/cni/conf.d"
+	ConfDir string
+
+	// CacheDir (optional) is the director in which CNI queries/results will be
+	// cached by the runtime. If not provided, defaults to "/var/lib/cni"
+	CacheDir string
+
+	// netNSPath is private to CNIConfiguration because we expect users to
+	// either provide the netNSPath via the Jailer config or allow the
+	// netns path to be autogenerated by us.
+	netNSPath string
+}
+
+func (cniConf CNIConfiguration) validate() error {
+	if cniConf.NetworkName == "" {
+		return errors.Errorf("must specify NetworkName in CNIConfiguration: %+v", cniConf)
+	}
+
+	return nil
+}
+
+func (cniConf *CNIConfiguration) setDefaults(machineID string) {
+	if cniConf.ContainerID == "" {
+		cniConf.ContainerID = machineID
+	}
+
+	if len(cniConf.BinPath) == 0 {
+		cniConf.BinPath = []string{defaultCNIBinDir}
+	}
+
+	if cniConf.ConfDir == "" {
+		cniConf.ConfDir = defaultCNIConfDir
+	}
+
+	if cniConf.CacheDir == "" {
+		cniConf.CacheDir = filepath.Join(defaultCNICacheDir, machineID)
+	}
+}
+
+func (cniConf CNIConfiguration) asCNIRuntimeConf() *libcni.RuntimeConf {
+	return &libcni.RuntimeConf{
+		ContainerID: cniConf.ContainerID,
+		NetNS:       cniConf.netNSPath,
+		IfName:      cniConf.IfName,
+		Args:        cniConf.Args,
+	}
+}
+
+func (cniConf CNIConfiguration) invokeCNI(ctx context.Context, logger *log.Entry) (*types.Result, error, []func() error) {
+	var cleanupFuncs []func() error
+
+	cniPlugin := libcni.NewCNIConfigWithCacheDir(cniConf.BinPath, cniConf.CacheDir, nil)
+
+	networkConf, err := libcni.LoadConfList(cniConf.ConfDir, cniConf.NetworkName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load CNI configuration from dir %q for network %q",
+			cniConf.ConfDir, cniConf.NetworkName), cleanupFuncs
+	}
+
+	runtimeConf := cniConf.asCNIRuntimeConf()
+
+	delNetworkFunc := func() error {
+		err := cniPlugin.DelNetworkList(ctx, networkConf, runtimeConf)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete CNI network list %q", cniConf.NetworkName)
+		}
+		return nil
+	}
+
+	// Try deleting the network in case it already exists, which can happen if there is
+	// a crash before cleanup from a previous invocations. If it doesn't exist,
+	// well-behaved CNI plugins should treat this as a no-op without returning an error.
+	// We can be reasonably sure any previous VM that was using this network is gone due
+	// to earlier validation that the VM's socket path does not already exist.
+	err = delNetworkFunc()
+	if err != nil {
+		logger.Errorf("failed to delete any pre-existing CNI network %+v: %v", cniConf, err)
```

Comment:
I may need to clarify the comment and update this.

Well-behaved plugins will not return any error from their delete method if the network doesn't exist, in which case `err` here will be nil. What's happening here is that they can encounter some other actual problem in the process of deleting things and actually fail to delete something.

Now that I've given this a second thought, I actually don't think this should be a log at all, it should return the error since an error may mean we are trying to create a network that was only partially deleted. I will update to do that.